src/Microsoft.AspNet.Authentication.OpenIdConnect/OpenIdConnectAuthenticationHandler.cs

```diff
@@ -217,14 +265,50 @@ public OpenIdConnectAuthenticationHandler(HttpClient backchannel)
 
             message.State = Options.StateDataFormat.Protect(properties);
 
-            var redirectUri = message.CreateAuthenticationRequestUrl();
-            if (!Uri.IsWellFormedUriString(redirectUri, UriKind.Absolute))
+            if (Options.AuthenticationMethod == OpenIdConnectAuthenticationMethod.RedirectGet)
+            {
+                var redirectUri = message.CreateAuthenticationRequestUrl();
+                if (!Uri.IsWellFormedUriString(redirectUri, UriKind.Absolute))
+                {
+                    Logger.LogWarning(Resources.OIDCH_0036_UriIsNotWellFormed, redirectUri);
+                }
+
+                Response.Redirect(redirectUri);
+
+                return true;
+            }
+            else if (Options.AuthenticationMethod == OpenIdConnectAuthenticationMethod.FormPost)
             {
-                Logger.LogWarning(Resources.OIDCH_0036_UriIsNotWellFormed, redirectUri);
+                var inputs = new StringBuilder();
+                foreach (var parameter in message.Parameters)
+                {
+                    var name = Options.HtmlEncoder.HtmlEncode(parameter.Key);
+                    var value = Options.HtmlEncoder.HtmlEncode(parameter.Value);
+
+                    var input = string.Format(CultureInfo.InvariantCulture, InputTagFormat, name, value);
+                    inputs.AppendLine(input);
+                }
+
+                var issuer = Options.HtmlEncoder.HtmlEncode(message.IssuerAddress);
```

Comment:
Wouldn't URL-encoding it make the resulting URL unusable (since it's used as the form action)?

https://github.com/aspnet/Security/pull/392#discussion_r36568911